src/lib/snapp.ts

```diff
@@ -292,31 +373,59 @@ export class SmartContract {
     // TODO: think about how address should be passed in
     // if address is not provided, create a random one
     address ??= PrivateKey.random().toPublicKey();
-    let dummyInstance = new this(address);
-    let inductiveRules = (this._methods ?? []).map(
-      ({ methodName, witnessArgs }) => {
-        function main(transactionHash: Field) {
-          Mina.setCurrentTransaction({
-            sender: PrivateKey.random(), // TODO
-            parties: [],
-            nextPartyIndex: 0,
-            protocolState: ProtocolStatePredicate.ignoreAll(),
-          });
-          let witnesses = witnessArgs.map(emptyWitness); // TODO is this correct?
-          (dummyInstance[methodName] as any)(...witnesses); // call method in a way that `this` is defined
-          let hash = computeTransactionHash(Mina.currentTransaction);
-          console.log({ transactionHash, hash });
-          hash.assertEquals(transactionHash);
-          Mina.setCurrentTransaction(undefined);
-        }
-        return [0, methodName, main];
-      }
+    let instance = new this(address);
+
+    let rules = (this._methods ?? []).map(({ methodName, witnessArgs }) =>
+      picklesRuleFromFunction(
+        methodName,
+        (...args: unknown[]) => (instance[methodName] as any)(...args),
+        witnessArgs
+      )
+    );
+
+    let output = withContext(
+      {
+        self: Party.defaultParty(address),
+        protocolState: ProtocolStatePredicate.ignoreAll(),
+      },
+      () => picklesCompile(rules)
     );
-    let output = picklesCompile(inductiveRules);
-    console.log('output', output);
     return output;
   }
 
+  async prove(provers: any[], methodName: keyof this, args: unknown[]) {
+    let SnappClass = this.constructor as never as typeof SmartContract;
```

Comment:
do you need the `as never` in there to get this to compile? is this the idiomatic way to force a cast in typescript?

Reply:
I've seen `as never` it a couple times. `as any` would also work, not sure which is cleaner. It seems to be needed here, because `this.constructor` is just `Function`. What I need here is access the child class -- the class that extends SmartContract. I didn't find a type-safe way to do that.